### PR TITLE
dnsdist: Add Lua bindings to interact with dynamic blocks

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

This PR adds more methods to interact (list, insert, be notified of a new insertion) with dynamic blocks from regular and FFI Lua scripts. 

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Review by chbruyand — COMMENTED** [review]
Just in case this is not voluntary, you seem to introduce a change in naming convention by expanding "Dyn" to "Dynamic" in some places.

File: pdns/dnsdist-lua.cc [hunk 1]
```diff
@@ -1418,6 +1418,54 @@ static void setupLuaConfig(LuaContext& luaCtx, bool client, bool configCheck)
...
+      }
+      auto entry = blockPair.second;
+      if (g_defaultBPFFilter && entry.bpf) {
+        entry.blocks += g_defaultBPFFilter->getHits(requestor.getNetwork());
+      }
+      if (entry.action == DNSAction::Action::None) {
+        entry.action = g_dynBlockAction;
+      }
+      entries.emplace(requestor.toString(), std::move(entry));
+      }
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdist-lua.cc]: Fixed, thanks!

File: pdns/dnsdist-lua.cc [hunk 2]
```diff
@@ -1418,6 +1418,54 @@ static void setupLuaConfig(LuaContext& luaCtx, bool client, bool configCheck)
...
+    return entries;
+  });
+
+  luaCtx.writeFunction("getSMTDynamicBlocks", []() {
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdist-lua.cc:1447]: Correct, I'll change that one.

File: pdns/dnsdist-lua-inspection.cc [hunk 3]
```diff
@@ -860,6 +860,11 @@ void setupLuaInspection(LuaContext& luaCtx)
         group->setSuffixMatchRuleFFI(seconds, reason, blockDuration, action ? *action : DNSAction::Action::None, std::move(visitor));
       }
     });
+  luaCtx.registerFunction<void(std::shared_ptr<DynBlockRulesGroup>::*)(dnsdist_ffi_dynamic_block_inserted_hook)>("setNewBlockInsertedHook", [](std::shared_ptr<DynBlockRulesGroup>& group, dnsdist_ffi_dynamic_block_inserted_hook hook) {
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdist-lua-inspection.cc]: Done

File: pdns/dnsdist-console.cc [hunk 4]
```diff
@@ -471,6 +471,7 @@ const std::vector<ConsoleKeyword> g_consoleKeywords{
   { "addDNSCryptBind", true, "\"127.0.0.1:8443\", \"provider name\", \"/path/to/resolver.cert\", \"/path/to/resolver.key\", {reusePort=false, tcpFastOpenQueueSize=0, interface=\"\", cpus={}}", "listen to incoming DNSCrypt queries on 127.0.0.1 port 8443, with a provider name of `provider name`, using a resolver certificate and associated key stored respectively in the `resolver.cert` and `resolver.key` files. The fifth optional parameter is a table of parameters" },
   { "addDOHLocal", true, "addr, certFile, keyFile [, urls [, vars]]", "listen to incoming DNS over HTTPS queries on the specified address using the specified certificate and key. The last two parameters are tables" },
   { "addDOQLocal", true, "addr, certFile, keyFile [, vars]", "listen to incoming DNS over QUIC queries on the specified address using the specified certificate and key. The last parameter is a table" },
+  { "addDynamicBlock", true, "address, message[, action [, seconds [clientIPMask [, clientIPPortMask]]]]", "block the set of addresses with message `msg`, for `seconds` seconds (10 by default), applying `action` (default to the one set with `setDynBlocksAction()`)" },
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdist-console.cc]: Applied, thanks!

**Comment by rgacogne** [comment]
> Just in case this is not voluntary, you seem to introduce a change in naming convention by expanding "Dyn" to "Dynamic" in some places.

Yeah, I find `DynBlock` not a very good name in user-facing methods, so I went with the full "Dynamic" for the new API.

File: pdns/dnsdist-console.cc [hunk 5]
```diff
@@ -471,6 +471,7 @@ const std::vector<ConsoleKeyword> g_consoleKeywords{
   { "addDNSCryptBind", true, "\"127.0.0.1:8443\", \"provider name\", \"/path/to/resolver.cert\", \"/path/to/resolver.key\", {reusePort=false, tcpFastOpenQueueSize=0, interface=\"\", cpus={}}", "listen to incoming DNSCrypt queries on 127.0.0.1 port 8443, with a provider name of `provider name`, using a resolver certificate and associated key stored respectively in the `resolver.cert` and `resolver.key` files. The fifth optional parameter is a table of parameters" },
   { "addDOHLocal", true, "addr, certFile, keyFile [, urls [, vars]]", "listen to incoming DNS over HTTPS queries on the specified address using the specified certificate and key. The last two parameters are tables" },
   { "addDOQLocal", true, "addr, certFile, keyFile [, vars]", "listen to incoming DNS over QUIC queries on the specified address using the specified certificate and key. The last parameter is a table" },
+  { "addDynamicBlock", true, "address, message[, action [, seconds [, clientIPMask [, clientIPPortMask]]]]", "block the set of addresses with message `msg`, for `seconds` seconds (10 by default), applying `action` (default to the one set with `setDynBlocksAction()`)" },
```
Comment (chbruyand) [inline pdns/dnsdist-console.cc:474]: This is not a set of addresses but a single address. (Sorry I missed this one the first time)

Comment (rgacogne) [reply pdns/dnsdist-console.cc:474]: No worries, I'm the one who messed up in the first place! Thanks :)

Comment (rgacogne) [reply pdns/dnsdist-console.cc]: Fixed

**Review by chbruyand — APPROVED** [review]

**Pull request merged**